File: packages/wasmvm/wasmclient/ts/wasmclient/tests/wasmclient.test.ts
import {WasmClientContext, WasmClientService} from '../lib';
import * as testwasmlib from 'testwasmlib';
import {
    addressFromBytes,
    bytesFromString,
    bytesToString,
    chainIDFromBytes,
    chainIDToBytes,
    chainIDToString, hexDecode,
    requestIDFromBytes
} from 'wasmlib';
import {KeyPair} from '../lib/iscclient';

const MYSEED = '0xa580555e5b84a4b72bbca829b4085a4725941f3b3702525f36862762d76c21f3';
const WASPAPI = 'http://localhost:19090';

const params = [
    'Lala',
    'Trala',
    'Bar|Bar',
    'Bar~|~Bar',
    'Tilde~Tilde',
    'Tilde~~ Bar~/ Space~_',
];

class EventProcessor {
    name = '';

    sendClientEventsParam(ctx: WasmClientContext, name: string) {
        const f = testwasmlib.ScFuncs.triggerEvent(ctx);
        f.params.name().setValue(name);
        f.params.address().setValue(ctx.currentChainID().address());
        f.func.post();
        checkError(ctx);
    }

    async waitClientEventsParam(ctx: WasmClientContext, name: string) {
        await this.waitEvent(ctx, 10000);
        checkError(ctx);
        expect(name == this.name).toBeTruthy();
        this.name = '';
    }

    private async waitEvent(ctx: WasmClientContext, msec: number): Promise<void> {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const self = this;
        return new Promise(function (resolve) {
            setTimeout(function () {
                if (self.name != '' || ctx.Err != null) {
                    resolve();
                } else if (msec <= 0) {
                    ctx.Err = 'event wait timeout';
                    resolve();
                } else {
                    self.waitEvent(ctx, msec - 100).then(resolve);
                }
            }, 100);
        });
    }
}

function checkError(ctx: WasmClientContext) {
    if (ctx.Err != null) {
        console.log('ERROR: ' + ctx.Err);
    }
    expect(ctx.Err == null).toBeTruthy();
}

let svc: WasmClientService;

function setupClient() {
    if (!svc) {
        svc = new WasmClientService(WASPAPI);
    }
    expect (svc.isHealthy()).toBeTruthy();
    const err = svc.setDefaultChainID();
    expect(err == null).toBeTruthy();
    const ctx = new WasmClientContext(svc, 'testwasmlib');
    ctx.signRequests(KeyPair.fromSubSeed(bytesFromString(MYSEED), 0));
    checkError(ctx);
    return ctx;
}

describe('keypair tests', function () {
    const mySeed = bytesFromString(MYSEED);
    it('construct proper sub-seed 0', () => {
        const subSeed = KeyPair.subSeed(mySeed, 0);
        console.log('Seed: ' + bytesToString(subSeed));
        expect(bytesToString(subSeed) == '0x65c0583f4d507edf6373e4bad8a649f2793bdf619a7a8e69efbebc8f6986fcbf').toBeTruthy();
    });

    it('construct proper sub-seed 1', () => {
        const subSeed = KeyPair.subSeed(mySeed, 1);
        console.log('Seed: ' + bytesToString(subSeed));
        expect(bytesToString(subSeed) == '0x8e80478dda48a3141e349ceac409ab9a4c742452c4e7e708d36fcb12b72b59d5').toBeTruthy();
    });

    it('should construct a proper pair', () => {
        const pair = new KeyPair(mySeed);
        console.log('Publ: ' + bytesToString(pair.publicKey));
        console.log('Priv: ' + bytesToString(pair.privateKey));
        expect(bytesToString(pair.publicKey) == '0x30adc0bd555d56ed51895528e47dcb403e36e0026fe49b6ae59e9adcea5f9a87').toBeTruthy();
        expect(bytesToString(pair.privateKey.slice(0, 32)) == '0xa580555e5b84a4b72bbca829b4085a4725941f3b3702525f36862762d76c21f3').toBeTruthy();
    });

    it('should construct sub-seed pair 0', () => {
        const pair = KeyPair.fromSubSeed(mySeed, 0);
        console.log('Publ: ' + bytesToString(pair.publicKey));
        console.log('Priv: ' + bytesToString(pair.privateKey));
        expect(bytesToString(pair.publicKey) == '0x80c6204f1aa7eb3a3c15f657d22aeec7f53f936a45d21f0a3bb2fb5de7fe002f').toBeTruthy();
        expect(bytesToString(pair.privateKey.slice(0, 32)) == '0x65c0583f4d507edf6373e4bad8a649f2793bdf619a7a8e69efbebc8f6986fcbf').toBeTruthy();
    });

    it('should construct sub-seed pair 1', () => {
        const pair = KeyPair.fromSubSeed(mySeed, 1);
        console.log('Publ: ' + bytesToString(pair.publicKey));
        console.log('Priv: ' + bytesToString(pair.privateKey));
        expect(bytesToString(pair.publicKey) == '0xfa482d65acb90e55dba4724886c79e1df663d3f95813a6674033504b89ffe7cd').toBeTruthy();
        expect(bytesToString(pair.privateKey.slice(0, 32)) == '0x8e80478dda48a3141e349ceac409ab9a4c742452c4e7e708d36fcb12b72b59d5').toBeTruthy();
    });

    it('should sign and verify', () => {
        const pair = new KeyPair(mySeed);
        const signedSeed = pair.sign(mySeed);
        console.log('Seed: ' + bytesToString(mySeed));
        console.log('Sign: ' + bytesToString(signedSeed));
        expect(bytesToString(signedSeed) == '0xa9571cc0c8612a63feaa325372a33c2f4ff6c414def18eb85ce4afe9b7cf01b84dba089278ca992e76fad8a50a76e3bf157216c445a404dc9e0424c250640906').toBeTruthy();
        expect(pair.verify(mySeed, signedSeed)).toBeTruthy();
    });
});

describe('ETH address strings', function () {
    it('should be proper ETH address string', () => {
        const svc = new WasmClientService(WASPAPI);
        svc.setCurrentChainID('atoi1ppp52dzsr6m2tle27v87e409n36xfcva3uld6lm093f0jgz2xng82pmf3yl');
        const strAddress = '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c';
        const address = addressFromBytes(hexDecode(strAddress));
        const ethAddress = address.toString();
        expect(strAddress == ethAddress).toBeTruthy();
    });
});

describe('wasmclient', function () {
    describe('call() view', function () {
        it('should call through web API', () => {
            const ctx = setupClient();

            const v = testwasmlib.ScFuncs.getRandom(ctx);
            v.func.call();
            checkError(ctx);
            const rnd = v.results.random().value();
            console.log('Rnd: ' + rnd);
            expect(rnd != 0n).toBeTruthy();
        });
    });

    describe('error handling', function () {
        it('should generate errors', () => {
            let ctx = setupClient();

            // missing mandatory string parameter
            const v = testwasmlib.ScFuncs.checkString(ctx);
            v.func.call();
            expect(ctx.Err != null).toBeTruthy();
            console.log('Error: ' + ctx.Err);

            // // wait for nonexisting request id (time out)
            // ctx.waitRequestID(requestIDFromBytes(null));
            // expect(ctx.Err != null).toBeTruthy();
            // console.log('Error: ' + ctx.Err);

            // sign with wrong wallet
            ctx.signRequests(KeyPair.fromSubSeed(bytesFromString(MYSEED), 1));
            const f = testwasmlib.ScFuncs.random(ctx);
            f.func.post();
            expect(ctx.Err != null).toBeTruthy();
            console.log('Error: ' + ctx.Err);

            // wait for request on wrong chain
            const chain_bytes = chainIDToBytes(ctx.currentChainID());
            chain_bytes[2]++;
            const badChainID = chainIDToString(chainIDFromBytes(chain_bytes));

            const svc = new WasmClientService(WASPAPI);
            ctx.Err = svc.setCurrentChainID(badChainID);
            checkError(ctx);
            ctx = new WasmClientContext(svc, 'testwasmlib');
            ctx.signRequests(KeyPair.fromSubSeed(bytesFromString(MYSEED), 0));
            ctx.waitRequestID(requestIDFromBytes(null));
            expect(ctx.Err != null).toBeTruthy();
            console.log('Error: ' + ctx.Err);
        });
    });

    describe('post() func request', function () {
        it('should post through web API', () => {
            const ctx = setupClient();

            const f = testwasmlib.ScFuncs.random(ctx);
            f.func.post();
            checkError(ctx);

            ctx.waitRequest();
            checkError(ctx);

            const v = testwasmlib.ScFuncs.getRandom(ctx);
            v.func.call();
            checkError(ctx);
            const rnd = v.results.random().value();
            console.log('Rnd: ' + rnd);
            expect(rnd != 0n).toBeTruthy();
        });
    });

    describe('event handling', function () {
        jest.setTimeout(20000);
        it('should receive multiple events', async () => {
            const ctx = setupClient();

            const events = new testwasmlib.TestWasmLibEventHandlers();
            const proc = new EventProcessor();
            events.onTestWasmLibTest((e) => {
                console.log(e.name);
                proc.name = e.name;
            });
            ctx.register(events);

            for (const param of params) {
                proc.sendClientEventsParam(ctx, param);
                await proc.waitClientEventsParam(ctx, param);
            }

            ctx.unregister(events.id());
            checkError(ctx);
        });
    });
});

File: packages/wasmvm/wasmclient/tests/wasmclient_test.rs
use std::sync::{Arc, Mutex};

use wasmlib::{address_from_bytes, chain_id_from_bytes, chain_id_to_bytes, chain_id_to_string, hex_decode, IEventHandlers, request_id_from_bytes};

use wasmclient::{self, iscclient::keypair, wasmclientcontext::*, wasmclientservice::*};

const MYSEED: &str = "0xa580555e5b84a4b72bbca829b4085a4725941f3b3702525f36862762d76c21f3";
const WASP_API: &str = "http://localhost:19090";

const PARAMS: &[&str] = &[
    "Lala",
    "Trala",
    "Bar|Bar",
    "Bar~|~Bar",
    "Tilde~Tilde",
    "Tilde~~ Bar~/ Space~_",
];

struct EventProcessor {
    name: Arc<Mutex<String>>,
}

impl EventProcessor {
    fn new() -> EventProcessor {
        EventProcessor {
            name: Arc::new(Mutex::new(String::new())),
        }
    }

    fn send_client_events_param(&self, ctx: &WasmClientContext, param: &str) {
        let f = testwasmlib::ScFuncs::trigger_event(ctx);
        f.params.name().set_value(param);
        f.params.address().set_value(&ctx.current_chain_id().address());
        f.func.post();
        check_error(ctx);
    }

    fn wait_client_events_param(&self, ctx: &WasmClientContext, param: &str) {
        for _ in 0..100 {
            {
                let name = self.name.lock().unwrap();
                if (*name).len() != 0 || ctx.err().is_err() {
                    break;
                }
            }
            std::thread::sleep(std::time::Duration::from_millis(100));
        }
        check_error(ctx);

        let mut name = self.name.lock().unwrap();
        assert_eq!(*name, param);
        name.clear();
    }
}

fn check_error(ctx: &WasmClientContext) {
    if let Err(e) = ctx.err() {
        println!("err: {}", e);
        assert!(false);
    }
}

fn setup_client() -> WasmClientContext {
    let mut svc = WasmClientService::new(WASP_API);
    assert!(svc.is_healthy());
    svc.set_default_chain_id().unwrap();
    let mut ctx = WasmClientContext::new(Arc::new(svc), "testwasmlib");
    let wallet = keypair::KeyPair::from_sub_seed(
        &wasmlib::bytes_from_string(MYSEED),
        0,
    );
    ctx.sign_requests(&wallet);
    check_error(&ctx);
    return ctx;
}

#[test]
fn sub_seeds() {
    let mut svc = WasmClientService::new(WASP_API);
    assert!(svc.is_healthy());
    svc.set_default_chain_id().unwrap();
    let my_seed = wasmlib::bytes_from_string(MYSEED);
    let mut sub_seed = keypair::KeyPair::sub_seed(&my_seed,0);
    let mut address = wasmlib::bytes_to_string(&sub_seed);
    assert_eq!(address, "0x65c0583f4d507edf6373e4bad8a649f2793bdf619a7a8e69efbebc8f6986fcbf");
    sub_seed = keypair::KeyPair::sub_seed(&my_seed,1);
    address = wasmlib::bytes_to_string(&sub_seed);
    assert_eq!(address, "0x8e80478dda48a3141e349ceac409ab9a4c742452c4e7e708d36fcb12b72b59d5");
}

#[test]
fn eth_address() {
    let mut svc = WasmClientService::new(WASP_API);
    svc.set_current_chain_id("atoi1ppp52dzsr6m2tle27v87e409n36xfcva3uld6lm093f0jgz2xng82pmf3yl").unwrap();
    let str_address = "0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c";
    let address = address_from_bytes(&hex_decode(str_address));
    let eth_address = address.to_string();
    assert_eq!(str_address, eth_address);
}

#[test]
fn call_view() {
    let ctx = setup_client();
    let v = testwasmlib::ScFuncs::get_random(&ctx);
    v.func.call();
    check_error(&ctx);
    let rnd = v.results.random().value();
    println!("rnd: {}", rnd);
    assert_ne!(rnd, 0);
}

#[test]
fn error_handling() {
    let mut ctx = setup_client();

    // missing mandatory string parameter
    let v = testwasmlib::ScFuncs::check_string(&ctx);
    v.func.call();
    assert!(ctx.err().is_err());
    println!("err: {}", ctx.err().err().unwrap());

    // // wait for nonexisting request id (time out)
    // ctx.wait_request_id(&request_id_from_bytes(&[]));
    // assert!(ctx.err().is_err());
    // println!("err: {}", ctx.err().err().unwrap());

    // sign with wrong wallet
    ctx.sign_requests(&keypair::KeyPair::from_sub_seed(
        &wasmlib::bytes_from_string(MYSEED),
        1,
    ));
    let f = testwasmlib::ScFuncs::random(&ctx);
    f.func.post();
    assert!(ctx.err().is_err());
    println!("err: {}", ctx.err().err().unwrap());

    let mut chain_bytes = chain_id_to_bytes(&ctx.current_chain_id());
    chain_bytes[2] += 1;
    let bad_chain_id = chain_id_to_string(&chain_id_from_bytes(&chain_bytes));

    // wait for request on wrong chain
    let mut svc = WasmClientService::new(WASP_API);
    svc.set_current_chain_id(&bad_chain_id).unwrap();
    let mut ctx = WasmClientContext::new(Arc::new(svc), "testwasmlib");
    ctx.sign_requests(&keypair::KeyPair::from_sub_seed(
        &wasmlib::bytes_from_string(MYSEED),
        0,
    ));
    ctx.wait_request_id(&request_id_from_bytes(&[]));
    assert!(ctx.err().is_err());
    println!("err: {}", ctx.err().err().unwrap());
}

#[test]
fn post_func_request() {
    let ctx = setup_client();
    let f = testwasmlib::ScFuncs::random(&ctx);
    f.func.post();
    check_error(&ctx);

    ctx.wait_request();
    check_error(&ctx);

    let v = testwasmlib::ScFuncs::get_random(&ctx);
    v.func.call();
    check_error(&ctx);
    let rnd = v.results.random().value();
    println!("rnd: {}", rnd);
    assert_ne!(rnd, 0);
}

#[test]
fn event_handling() {
    let ctx = setup_client();
    let mut events = testwasmlib::TestWasmLibEventHandlers::new();

    let proc = EventProcessor::new();
    {
        let name = proc.name.clone();
        events.on_test_wasm_lib_test(move |e| {
            println!("{}", e.name.to_string());
            let mut name = name.lock().unwrap();
            *name = e.name.clone();
        });
    }
    let events_id = events.id();
    ctx.register(Box::new(events));
    check_error(&ctx);

    for param in PARAMS {
        proc.send_client_events_param(&ctx, &param);
        proc.wait_client_events_param(&ctx, &param);
    }

    ctx.unregister(events_id);
    check_error(&ctx);
}
